### PR TITLE
P3-307 Add improved Elementor compatibility

### PR DESCRIPTION
**Pull request by jcomack**
## Context

* Improves compatibility with Elementor and the Rewrite and Republish feature.

## Summary

This PR can be summarized in the following changelog entry:

* Improves compatibility with Elementor and the Rewrite and Republish feature.

## Relevant technical choices:

*

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

**Please note that these instructions also apply when using the Classic Editor**

#### Prerequisites
* Checkout this branch and build accordingly (`composer install`, `yarn` and `grunt build`)

#### Creating an Elementor R&R copy from a regular post
* Ensure you have both Yoast SEO and Elementor installed and activated
* Go to Posts and click on Rewrite & Republish on your desired, non-Elementor post.
* Click on Edit with Elementor
* Make some changes to the content and click on the little cog in Elementors sidebar (bottom left by default)
* See that you can't change post status (field isn't there)
* Edit the featured image
* Click the Yoast SEO button at the top of the sidebar and edit the keyphrase. You should see a message regarding saving changes to the SEO settiings. _If you wait long enough, Elementor will autosave these changes to the SEO settings for you. Alternatively, you can click the arrow next to the Republish button and Save as Draft first _
* Click the arrow next to the green Republish button in the bottom left. See that the Save as Template feature isn't there.
* Click Republish. You should receive a message that republishing was successful and be redirected to the original post.
* _In case you didn't wait for Elementor to auto-save the changes made in the SEO settings, nor did you click Save as Draft, you should see that the keyphrase field remains empty (or contains an old keyphrase)_
* Verify that the other changes you made in the copy, have been copied over to the post you're currently viewing

#### Creating an Elementor R&R copy from an Elementor post
* Ensure you have both Yoast SEO and Elementor installed and activated
* Go to Posts and click on Rewrite & Republish on a pre-existing Elementor post (or create a new one first)
* Click on Edit with Elementor
* Make some changes to the content and click on the little cog in Elementors sidebar (bottom left by default)
* See that you can't change post status (field isn't there)
* Edit the featured image
* Click the Yoast SEO button at the top of the sidebar and edit the keyphrase. You should see a message regarding saving changes to the SEO settiings. _If you wait long enough, Elementor will autosave these changes to the SEO settings for you. Alternatively, you can click the arrow next to the Republish button and Save as Draft first _
* Click the arrow next to the green Republish bottom in the bottom left. See that the Save as Template feature isn't there.
* Click Republish. You should receive a message that republishing was successful and be redirected to the original post.
* _In case you didn't wait for Elementor to auto-save the changes made in the SEO settings, nor did you click Save as Draft, you should see that the keyphrase field remains empty (or contains an old keyphrase)_
* Verify that the other changes you made in the copy, have been copied over to the post you're currently viewing

### Test instructions for QA when the code is in the RC

* [x] QA should use the same steps as above. (Skip the Prerequisites section)

QA can test this PR by following these steps:

*

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes

* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Documentation

* [ ] I have written documentation for this change.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [x] I have added unittests to verify the code works as intended

Fixes [P3-307]

### CODE REVIEW

File: js/src/duplicate-post-elementor.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
+/* global $e */
+
+/**
+ * Hooks into Element on initialization.
```
Comment (karlijnbok) [inline js/src/duplicate-post-elementor.js:4]: Element or Elementor?

File: src/class-post-republisher.php [hunk 2]
```diff
@@ -106,6 +106,11 @@ public function register_post_statuses() {
 	 * @return array An array of slashed, sanitized, and processed attachment post data.
 	 */
 	public function change_post_copy_status( $data, $postarr ) {
+		if ( ! \array_key_exists( 'ID', $postarr )
+		|| empty( $postarr['ID'] ) ) {
```
Comment (karlijnbok) [inline src/class-post-republisher.php:110]: Is there a specific reason this is not on one line?

**Comment by karlijnbok** [comment]
I've committed my CR comments, after having discussed it with @jcomack.   

CR: 👍

File: js/src/duplicate-post-elementor.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+        /**
+         * Applies the redirect if the condtions are met.
+         *
+         * @param {Object} args The arguments to use.
+         *
+         * @returns {void}
+         */
+        apply( args ) {
+            if ( args.status === "publish" && duplicatePost.originalEditURL ) {
```
Comment (afercia) [inline js/src/duplicate-post-elementor.js:54]: `duplicatePost` is a global, it should be added in the comment at the top of the file.

File: js/src/duplicate-post-elementor.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+ * Removes the Save as Template option for Rewrite and Republish copy.
+ *
+ * @returns {void}
+ */
+function duplicatePostRemoveSaveTemplate() {
+    if ( duplicatePost.rewriting === "0" ) {
+        return;
+    }
+
+    elementor
```
Comment (afercia) [inline js/src/duplicate-post-elementor.js:73]: To my understanding, `elementor` is a global and should be added in the comment at the top of the file.

File: js/src/duplicate-post-elementor.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+// Wait on `window.elementor`.
+jQuery( window ).on( "elementor:init", () => {
+    // Wait on Elementor app to have started.
+    window.elementor.on( "panel:init", () => {
+        duplicatePostOnElementorInitialize();
+        duplicatePostRemoveSaveTemplate();
+    } );
+} );
```
Comment (afercia) [inline js/src/duplicate-post-elementor.js:89]: Indentation across the whole file is made with spaces. It should be made with tab characters.

File: js/src/duplicate-post-elementor.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    if ( duplicatePost.rewriting === "0" ) {
+        return;
+    }
+
+    elementor
+        .getPanelView()
+        .footer
+        .currentView
+        .removeSubMenuItem( 'saver-options', {
+            name: 'save-template',
```
Comment (afercia) [inline js/src/duplicate-post-elementor.js:78]: In the two lines above: we should use double quotes.

**Comment by afercia** [comment]
CR 2 and Acceptance 👍

**Pull request merged**